### PR TITLE
[Perf] Optimize Higgs Audio V3 request-aware decode pipeline

### PR DESCRIPTION
**Pull request by Sy0307**
## Purpose

Higgs Audio V3 high-concurrency decoding previously serialized each GPU step with D2H synchronization, per-request audio postprocessing, and scheduler/output construction. On A100 c64 this left large host-side bubbles and made vLLM-Omni substantially slower than the matched SGL-Omni path.

This PR:

- introduces request-owned decode and sampling state so batch shrink, reorder, finish, cancel, and row reuse do not leak state across requests;
- overlaps step N audio postprocessing with step N+1 launch using request-aware pending handles and double-buffered pinned-host staging;
- adds compiled tensor audio-state updates and a logits-free direct-audio path, while retaining the stock sampler whenever logits-dependent semantics are requested;
- resolves one immutable effective seed per request and binds seed/step state by request ID;
- aligns the high-throughput profile with the validated A100 c64 configuration;
- scopes request-ID snapshots behind `requires_request_ids_for_decode_state`, so the generic runner does not require unrelated models to consume `req_ids`.

## Test Plan

- [x] Run Higgs V3 and runner unit tests on NVIDIA A100.
- [x] Run Seed-TTS unique500, concurrency 64, single-GPU/single-replica throughput tests.
- [x] Compare against latest SGL-Omni with prefix/Radix cache disabled.
- [x] Run Seed-TTS unique64 UTMOS/WER quality validation.
- [x] Verify request reorder, shrink, finish/cancel, row reuse, seed ownership, logits fallback, and request-ID capability behavior.
- [x] Run Ruff, formatting checks on changed files, compileall, and git diff checks.

**vLLM Version:** 0.25.0

**vLLM-Omni Commit:** 7b42a84aaa6358010e70e9730ce4e79c0836cab3

### Correctness

Final A100 targeted suite:

```text
181 passed, 19 warnings
```

The request-ID capability regression was verified TDD-style: the old generic runner failed with `unexpected keyword argument 'req_ids'`; the capability-gated implementation passes both the generic-model and Higgs snapshot cases.

### Performance

Test setup: NVIDIA A100 80GB, Seed-TTS unique500, concurrency 64, one GPU, one replica, 500 unique prompts, `max_new_tokens=256`, and prefix/Radix cache disabled.

#### PR vs main baseline

| Metric | Main baseline | PR | Absolute diff | Relative diff |
| --- | ---: | ---: | ---: | ---: |
| Audio throughput | 54.9580 audio-s/s | 95.2728 audio-s/s | +40.3148 audio-s/s | **+73.36%** |
| Request throughput | 12.9300 QPS | 22.5510 QPS | +9.6210 QPS | **+74.41%** |

The main result is the matched old-main A100 c64 baseline. The PR result is the mean of three A100 c64 runs on the final request-aware branch.

#### Throughput gain attribution (conditional A/B)

The end-to-end PR-vs-main result is **+73.36% audio-s/s** and **+74.41% QPS**. The A/B results below answer narrower, conditional questions; they are not additive components of the 73.36% total:

| Conditional A/B | Fixed condition | Before | After | Relative diff |
| --- | --- | ---: | ---: | ---: |
| Steady-state codec chunk: 25 → 75 frames | Pending overlap on | 77.050 audio-s/s | 95.819 audio-s/s | **+24.36%** |
| Request-aware pending overlap: off → on | Chunk size 75 | 82.119 audio-s/s | 95.819 audio-s/s | **+16.68%** |
| Audio state update: legacy → compiled | Chunk size 75, overlap on | 92.545 audio-s/s | 95.819 audio-s/s | **+3.54%** |

The denominators differ: the chunk effect was measured with overlap enabled, while the overlap effect was measured only at chunk size 75. Therefore, adding the three percentages (44.58%) or multiplying them (50.24%) is not a valid decomposition of the PR-vs-main gain.

The total PR-vs-main improvement also contains the rest of the optimized path—logits-free text-logit skipping, c64 CUDA Graph coverage, request-owned state and sampling, high-throughput configuration alignment—and interactions between those changes. An exact factorial attribution would additionally require the missing chunk-25/pending-off cell under the same candidate code and configuration; old main is not that cell because it differs in multiple other paths.

The chunk change keeps the initial chunk at one frame and only increases the steady-state chunk to 75 frames, amortizing Stage1 codec invocations, connector traffic, and audio-chunk publication overhead. Pending overlap instead hides the previous step's D2H/event wait and request routing behind the next decode step.

#### PR vs latest SGL-Omni

| Metric | Latest SGL-Omni | PR | PR relative diff |
| --- | ---: | ---: | ---: |
| Audio throughput | 92.4891 audio-s/s | 95.2728 audio-s/s | **+3.01%** |
| Request throughput | 21.7110 QPS | 22.5510 QPS | **+3.87%** |

Both PR and SGL runs completed 500/500 requests per run. The SGL comparison used the latest validated SGL-Omni environment with the same c64 single-replica shape and cache disabled.

### Quality

Seed-TTS unique64:

| Implementation | UTMOS | WER |
| --- | ---: | ---: |
| Main reference | 3.7689 | 0.0558 |
| PR | 4.3761 | 0.0580 |

All 64 PR outputs were finite with normal amplitude. The request-owned seed lifecycle is deterministic, but full c64 byte-for-byte E2E determinism is not claimed because dynamic batch shapes and GPU kernel numerics can still change the autoregressive trajectory.

**BEFORE SUBMITTING:** read [CONTRIBUTING.md](https://github.com/vllm-project/vllm-omni/blob/main/CONTRIBUTING.md) and run the [precheck-pr skill](https://github.com/vllm-project/vllm-omni/blob/main/.claude/skills/precheck-pr/SKILL.md) with the code agent for a self-check against project conventions.
(anything written below this line will be removed by GitHub Actions)

### CODE REVIEW

**Comment by Sy0307** [comment]
cc @linyueqian @hsliuustc0106

File: vllm_omni/model_executor/models/higgs_audio_v3/higgs_audio_v3_talker.py [hunk 1]
```diff
@@ -1334,36 +2193,368 @@ def _update_delay_state_batched(
...
+            # Stateless request RNG: each draw is keyed by
+            # (request seed, AR step, codebook). This is batch-order invariant
+            # and pure tensor code, unlike a Python loop over torch.Generator.
+            modulus = 2_147_483_647
+            seeds = torch.remainder(row_seeds.to(device=probs.device, dtype=torch.long), modulus)
+            steps = row_steps.to(device=probs.device, dtype=torch.long)
+            codebooks = torch.arange(num_codebooks, device=probs.device, dtype=torch.long).unsqueeze(0)
+            positions = steps.unsqueeze(1) * num_codebooks + codebooks + 1
+            state = torch.remainder(seeds.unsqueeze(1) + positions * 1_103_515_245, modulus)
+            state = torch.remainder(torch.bitwise_xor(state, state >> 16) * 48_271, modulus)
```
Comment (linyueqian) [inline vllm_omni/model_executor/models/higgs_audio_v3/higgs_audio_v3_talker.py:2411]: [suggestion] This mixer is a single LCG multiply plus one xorshift, and it leaves visible structure: consecutive-step draws for the same codebook correlate at about 0.25, and two seeds that differ by 1 produce per-draw uniforms that agree within ~2e-5 (correlation 0.9999; seeds only decorrelate at distances around 1e5). In end-to-end runs adjacent-seed requests still diverge, because one flipped token reroutes the AR trajectory (I checked 6 adjacent-seed pairs, all produced different audio), so this is not blocking. Still, a stronger stateless finalizer (e.g. a SplitMix64-style xor-shift-multiply sequence over (seed, step, codebook)) would remove both artifacts at no measurable cost and keep the batch-order invariance this design gets right.

File: vllm_omni/model_executor/models/higgs_audio_v3/higgs_audio_v3_talker.py [hunk 2]
```diff
@@ -41,7 +45,96 @@
...
+@torch.library.custom_op(
+    "vllm_omni::higgs_flashinfer_top_k_renorm",
+    mutates_args=("row_states",),
+)
+def _higgs_flashinfer_top_k_renorm_op(
+    probs: torch.Tensor,
+    row_states: torch.Tensor,
+) -> torch.Tensor:
+    module = _get_higgs_flashinfer_sampling_module()
+    return module.top_k_renorm_probs(probs, None, 50, row_states)
```
Comment (linyueqian) [inline vllm_omni/model_executor/models/higgs_audio_v3/higgs_audio_v3_talker.py:76]: [suggestion] The op body hardcodes top_k=50 here (and top_p=0.95 in the top-p op below) while the torch path re-derives the same values inside _sample_audio_codes_tensor. If these ever become configurable the two sampler paths will silently drift. Consider hoisting both into shared module-level constants.

File: vllm_omni/model_executor/models/higgs_audio_v3/higgs_audio_v3_talker.py [hunk 3]
```diff
@@ -62,6 +155,273 @@
...
+        self._in_use = [False, False]
+        self._cursor = 0
+        self._condition = threading.Condition()
+
+    def acquire(self, num_rows: int) -> _HiggsAudioHostStagingLease:
+        num_rows = max(0, int(num_rows))
+        with self._condition:
+            slot = self._cursor
+            self._cursor ^= 1
+            while self._in_use[slot]:
```
Comment (linyueqian) [inline vllm_omni/model_executor/models/higgs_audio_v3/higgs_audio_v3_talker.py:274]: [suggestion] acquire() blocks forever if a taken pending handle is ever dropped without resolve(); an exception between take_pending_omni_postprocess and the background resolve would leak the lease, and a second leak deadlocks the decode loop. Client aborts are handled fine (I exercised four mid-stream disconnects without a wedge), so this is theoretical, but a generous wait timeout that raises a loud error would turn a silent engine hang into a diagnosable failure.

File: vllm_omni/model_executor/models/higgs_audio_v3/higgs_audio_v3_talker.py [hunk 4]
```diff
@@ -1334,36 +2193,368 @@ def _update_delay_state_batched(
...
+            rows = max(num_rows, self._mlp_cudagraph_max_batch)
+            buf = torch.empty((rows, hidden_size), dtype=dtype, device=device)
+            self._audio_state_hidden_staging = buf
+        return buf[:num_rows]
+
+    def _get_audio_state_step_callable(self) -> Any:
+        if self._audio_state_step_mode != "compile":
+            return self._audio_state_step_from_hidden
+        if self._compiled_audio_state_step is None:
+            self._compiled_audio_state_step = torch.compile(
```
Comment (linyueqian) [inline vllm_omni/model_executor/models/higgs_audio_v3/higgs_audio_v3_talker.py:2256]: [suggestion] The first minute of traffic after server start pays this compile through live requests: my first bench rep after startup showed TTFP p99 of 3.6-6.1s versus a steady-state p99 under 900ms. Worth considering a startup warmup that runs the compiled state step across the capture bucket sizes before the server reports ready.

**Review by linyueqian — APPROVED** [review]
Verified this end to end on a single H20 141GB (vllm 0.25.0, head fc8ef8ca vs merge-base 0d8c6326, each side serving its own shipped high_throughput profile, Seed-TTS EN text, 256 unique prompts per rep, cold rep discarded).

Correctness: the 181-test unit suite passes on GPU. Serving battery on the auto-discovered profile is green: unseeded repeats differ (the per-request effective seed plumbing works), seeded repeats are bit-identical, c=8 concurrent requests transcribe with zero cross-request contamination, SSE streaming works, and four genuine mid-stream client aborts did not wedge the pending staging pool. Request cancel, reorder, and finish behavior all held up under concurrent load.

Perf (audio-s/s, warm reps): c=1 goes 2.51 to 5.33 (+112%), c=16 goes 35 to 75 (+115%, mean TTFP 233ms to 150ms), c=64 goes 76.9 (best baseline rep, spread 43-77) to 131.5/132.1 (+71%, tight across reps). Quality on the same 128 prompts and Whisper protocol: mean WER 0.0947 to 0.0134, consistent with the UTMOS gain reported in the PR body (larger codec chunks cut boundary artifacts). The A100 numbers reproduce directionally on H20 and are exceeded at low and mid concurrency.

A few non-blocking suggestions inline; none of them block merge. Nice work, this closes the eager dispatch bottleneck that kept this model behind at low concurrency.
